libs/shared/shared_common.c

```diff
@@ -648,3 +648,77 @@ ebpf_duplicate_program_data(
 
     EBPF_RETURN_RESULT(result);
 }
+
+ebpf_result_t
+ebpf_canonicalize_path(_Out_writes_(output_size) char* output, size_t output_size, _In_z_ const char* input)
+{
+    const int PREFIX_SIZE = 5; // Length of "BPF:\\".
+    strcpy_s(output, output_size, "BPF:\\");
+
+    // Add the BPF: prefix if not already present.
+    if (_strnicmp(input, "BPF:\\", PREFIX_SIZE) == 0 || _strnicmp(input, "BPF:/", PREFIX_SIZE) == 0) {
+        strcpy_s(output + PREFIX_SIZE, output_size - PREFIX_SIZE, input + PREFIX_SIZE);
+    } else {
+        strcpy_s(output + PREFIX_SIZE, output_size - PREFIX_SIZE, input);
+    }
+
+    for (int i = PREFIX_SIZE; output[i] != 0; i++) {
+        if (output[i] == '/') {
+            // Convert slashes to backslashes.
+            output[i] = '\\';
+        } else if (output[i] == ':') {
+            // Disallow ':' in mid-path.
+            return EBPF_INVALID_ARGUMENT;
+        }
+    }
+
+    // Remove other prefix aliases from the beginning.
+    if (_strnicmp(output + PREFIX_SIZE, "\\ebpf\\global\\", 13) == 0) {
+        char* next = output + PREFIX_SIZE + 13;
+        memmove(output + PREFIX_SIZE, next, strlen(next) + 1);
+    }
+
+    for (int i = PREFIX_SIZE - 1; output[i] != 0;) {
+        if (strncmp(output + i, "\\\\", 2) == 0) {
+            char* next = output + i + 2;
+            memmove(output + i + 1, next, strlen(next) + 1);
+
+            // Stay at this position for the next loop iteration.
+        } else if (strncmp(output + i, "\\.\\", 3) == 0) {
+            char* next = output + i + 3;
+            memmove(output + i + 1, next, strlen(next) + 1);
+        } else if (strncmp(output + i, "\\..\\", 4) == 0) {
+            int previous_index = i - 1;
+            for (; previous_index >= 4; previous_index--) {
```

Comment:
Why 4 as the boundary value?

Reply:
"BPF:" is length 4.

Reply:
Updated to be easier to follow (I hope).